src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -445,6 +445,23 @@ class ResourceService(
       _ <- AuditLogger.logAuditEventIO(samRequestContext, ResourceEvent(ResourceDeleted, resource))
     } yield ()
 
+  @throws(classOf[WorkbenchExceptionWithErrorReport]) // Necessary to make Mockito happy
+  def deleteResourceCascade(resource: FullyQualifiedResourceId, samRequestContext: SamRequestContext): IO[Unit] =
+    for {
+      _ <- checkNoChildren(resource, samRequestContext)
+      _ <- accessPolicyDAO.removePolicyGroupsInUse(resource, samRequestContext)
+
+      // remove from cloud first so a failure there does not leave sam in a bad state
+      _ <- cloudDeletePolicies(resource, samRequestContext)
+      _ <- deleteActionManagedIdentitiesForResource(resource, samRequestContext)
+
+      // leave a tomb stone if the resource type does not allow reuse
+      leaveTombStone = !resourceTypes(resource.resourceTypeName).reuseIds
+      _ <- accessPolicyDAO.deleteResource(resource, leaveTombStone, samRequestContext)
+
+      _ <- AuditLogger.logAuditEventIO(samRequestContext, ResourceEvent(ResourceDeleted, resource))
+    } yield ()
+
```

Comment:
there's a lot of copy-paste here between `deleteResource` and `deleteResourceCascade`. Can these be consolidated? Looks like the only difference is the invocation of `checkNoPoliciesInUse(resource, samRequestContext)` vs `accessPolicyDAO.removePolicyGroupsInUse(resource, samRequestContext)` in the second step of the for comprehension.

Reply:
Yep!  This was all just to prove the concept, definitely planning on cleaning it up.